File: src/java/tusan/nodes/base/function/ExecuteFunction.java
package net.tutla.tums.tusan.nodes.base.function;

import net.tutla.tums.tusan.Node;
import net.tutla.tums.tusan.Types;
import net.tutla.tums.tusan.Utils;
import net.tutla.tums.tusan.Variable;
import net.tutla.tums.tusan.interpreter.Interpreter;
import net.tutla.tums.tusan.lexer.Token;
import net.tutla.tums.tusan.lexer.TokenType;
import net.tutla.tums.tusan.nodes.expression.Expression;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class ExecuteFunction extends Node {
    public Object value;
    public FunctionRegistry originalFunction;
    public ExecuteFunction(Token token){
        super(token);
    }

    public ExecuteFunction create(){
        interpreter.expectTokenType(TokenType.LEFT_PAR);

        originalFunction = interpreter.data.funcs.get(token.value);
        String name = token.value;
        Interpreter functionInterpreter = originalFunction.interpreter.clone();
        List<FunctionParameter> parameters = originalFunction.parameters;

        HashMap<String, FunctionParameter> optionalParameters = new HashMap<>();
        Integer parameterPointer = 0;
        for (FunctionParameter parameter : parameters){ // checking the required ones
            if (parameter.required){
                Object val = new Expression(interpreter.nextToken()).create().value;

                if (parameter.type != Types.ANY){
                    Types valType = Utils.getTypeOfValue(val);
                    if (valType != parameter.type){
                        interpreter.error("TypeError", "Received type "+valType+" instead of "+parameter.type.name()+" in function {token.value} ", null);
                    }
                }
                functionInterpreter.data.vars.put(parameter.name, new Variable(parameter.name, val, new HashMap<>()));
            } else { // putting the optional ones into a hashmap
                optionalParameters.put(parameter.name, parameter);
            }
        }

        while (optionalParameters.containsKey(interpreter.getNextToken().value)){ // evaluating the optional ones
            FunctionParameter parameter = optionalParameters.get(interpreter.getNextToken().value);
            interpreter.expectTokenType(TokenType.EQUAL);
            Object val = new Expression(interpreter.nextToken()).create().value;
            if (parameter.type != Types.ANY){
                Types valType = Utils.getTypeOfValue(val);
                if (valType != parameter.type){
                    interpreter.error("TypeError", "Received type "+valType+" instead of "+parameter.type.name()+" in function {token.value} ", null);
                }
            }
            functionInterpreter.data.vars.put(parameter.name, new Variable(parameter.name, val, new HashMap<>()));
            optionalParameters.remove(interpreter.getNextToken().value); // the ones that remain will take the fallback value

            if (interpreter.getNextToken().type == TokenType.COMMA){
                interpreter.nextToken();
            }
        }

        for (FunctionParameter parameter : optionalParameters.values()){
            functionInterpreter.data.vars.put(parameter.name, new Variable(parameter.name, parameter.fallback, new HashMap<>()));
        }

        interpreter.expectTokenType(TokenType.RIGHT_PAR);
        functionInterpreter.changeTokensParent(functionInterpreter);
        functionInterpreter.isFunction = true;
        functionInterpreter.compile();

        value = functionInterpreter.returned;

        return this;
    }
}

File: src/python/src/tusan/nodes/base/function/ExecuteFunction.py
from tusan.Node import *

class ExecuteFunctionNode(Node):
    def __init__(self, token: Token):
        self.interpreter = token.interpreter
        self.token = token

    async def create(self):
        from tusk.nodes.expressions import ExpressionNode
        from tusk.variable import get_type_
        token = self.token
        self.name = token.value
        
        func = token.interpreter.data["funcs"][token.value] # [[], interpreter]
        func_name= token.value
        func_interpreter = func["interpreter"]
        func_params = func["params"]

        
        parased_params = []
        nparams = {}
        for param in func_params: # looping params (func[0] is the param list)
            if param["fallback"] == [None, 5678, "no chance"]: # i had no other idea to get around it so we're doing this list, so if the user by accident provides [None, 5678, "no chance"] they're cooked 💀
                e = self.interpreter.next_token() 

                val = (await ExpressionNode(e).create()).value
                if param["type"] != "ANY":
                    if (await get_type_(val)) == param["type"]:
                        parased_params.append([param["name"],val])
                    else:
                        self.interpreter.error("TypeError", f"Recieved type {(await get_type_(val))} instead of {param['type']} in function {token.value} ") 
                else:
                    parased_params.append([param["name"],val])
            else:
                nparams[param["name"]] = param
        while self.interpreter.get_next_token().value in nparams:
            if nparams[self.interpreter.get_next_token().value]["fallback"] != [None, 5678, "no chance"]:
                self.interpreter.error("SyntaxError",f"What is a required argument ({nparams[self.interpreter.get_next_token().value]['name']}) doing here?")
            else:
                self.interpreter.error("SyntaxError",f"Invalid paramater {self.interpreter.get_next_token().value} in function {token.value}")
            nparams.pop(self.interpreter.get_next_token().value)
        for i in nparams:
            parased_params.append([nparams[i]["name"],nparams[i]["fallback"]])
        

        for i in parased_params: func_interpreter.data["vars"][i[0]] = i[1]

        func_interpreter.data["funcs"] = self.interpreter.data["funcs"]
        self.value = await func_interpreter.compile()
        self.value = func_interpreter.return_value
        return self